syntax = "proto3";

option go_package = "./minio-service";

package transmission;

service Transmission {

    rpc CreateUser(User) returns (User){}
    rpc UpdateUser(User) returns (User){}
    rpc ReadUser(User) returns (User){}

    rpc CreateWorkspace(Workspace) returns (ID){}

    rpc CreateFile(File) returns (ID){}
    rpc GetFile(WorkspaceFile) returns (File){}
    rpc DeleteFile(File) returns (Status){}

    rpc CreateFolder(Folder) returns (Status){}
    rpc GetFolder(Folder) returns (files){}
    rpc DeleteFolder(Folder) returns (Status){}

}

message ID{
    string id = 1;
}

message files{
    repeated File files =1;
    int32 total = 2;
}

message User{
    optional string id =1;
    string login =2;
    string password =3;
    string workspace_id=4;
    string role=5; 
}

message Workspace{
    optional string id =1;
    string name =2;
    string owner =3;
}

message WorkspaceFile{
    string workspace_id=1;
    string path =2;
    bytes buffer =3;
}

message File{
    optional string id =1;
    string path =2;
    string workspace_id=3;
    optional bytes buffer=4;
}

message Folder{
    optional string id =1;
    string path =2;
    string workspace_id=3;
    int32 skip=4;
    int32 take=5;
}


message Binary {
    bytes binary = 2;
}

message Path {
    string path = 1;
}

message Status {
    bool status = 1;
}

message SendFile {
    Path path_to_new_file = 1;
    Binary binary_file = 2;
}

message TakeFile {
    optional Binary binary_file = 1;
    Status bool_status = 2;
}